src/server/scripts/poller.js

```diff
@@ -0,0 +1,274 @@
+      return false;
+    default:
+      // Bail, unexpected response.
+      throw new Error(
+        `${repositoryUrl} (${response.statusCode}): ${response.body.message}`);
+  }
+};
+
+
+// Extracts unique GH repository URLs from a given (parsed) snapcraft.yaml
+export function extractPartsToPoll(snapcraft_yaml) {
+  const parts = Object.values(snapcraft_yaml.parts || {});
+  const sourceParts = parts.map(
+    GitSourcePart.fromSnapcraftPart).filter(part => part != undefined);
+  return Array.from(new Set(sourceParts));
+}
+
+
+/// GitSourcePart encapsulates the relevant information from a snapcraft
+/// source part, and contains the methods to determine whether it's up to
+/// date.
+export class GitSourcePart {
+
+  constructor(repoUrl, branch, tag) {
+    if (repoUrl === undefined) {
+      throw new Error('Required parameter: repoUrl');
+    }
+    if (branch === undefined) {
+      branch = 'master';
+    }
+    this.repoUrl = repoUrl;
+    this.branch = branch;
+    this.tag = tag;
+  }
+
+  /** Extract a GitSourcePart from a snapcraft part definition.
+   *
+   * Returns a GitSourcePart instance if the source part meets the following
+   * criteria:
+   *
+   * - The source part has a source repository listed, and it's a github
+   *   hosted repository
+   */
+  static fromSnapcraftPart(part) {
+    // TODO: Warn if source-commit or source-subdir are set, since we don't
+    //       support these.
+    // TODO: Not sure if we can support setting tags _and_ branch in the same
+    //       part.
+    const gh_repo_prefix = conf.get('GITHUB_REPOSITORY_PREFIX');
+    if (part.source == undefined) {
+      logger.info('Skipping part with no source set.');
+    } else if (part.source.startsWith(gh_repo_prefix)) {
+      var sourceUrl = part['source'];
+      var sourceBranch = part['source-branch'];
+      var sourceTag = part['source-tag'];
+      // TODO: figure out tag support:
+      if (sourceTag) {
+        logger.info(
+          `Not checking ${sourceUrl} with tag ${sourceTag} since tags are not supported`);
+        return;
+      }
+      return new GitSourcePart(sourceUrl, sourceBranch, sourceTag);
+    } else {
+      logger.info(
+        `Not checking ${part.source} as only github repos are supported`);
+    }
+  }
+
+  /** Determine if the source part has changed since `last_polled_at`
+   *
+   */
+  async hasRepoChangedSince(last_polled_at, token) {
+    if (last_polled_at === undefined || !last_polled_at) {
+      throw new Error('`last_polled_at` must be given.');
+    }
+    const last_polled = moment(last_polled_at);
+    const since = last_polled.toISOString();
+    const { owner, name } = parseGitHubRepoUrl(this.repoUrl);
+
+    const options = {
+      token,
+      headers: {
+        'If-Modified-Since': last_polled.format('ddd, MM MMM YYYY HH:mm:ss [GMT]')
+      },
+      json: true
+    };
+
+    if (this.branch === 'master' && this.tag == undefined) {
```

Comment:
Right, I've dropped the commits API usage and it's now routing every change check via the branches API right after figuring out what is the 'default' branch.